// @progress/kendo-angular-dialog/dist/es/dialog.component.js
import { Component, ContentChild, EventEmitter, HostBinding, Input, ViewChild, Output, ElementRef, Renderer, trigger, state, style, transition, animate } from '@angular/core';
import { DialogActionsComponent } from './dialog-actions.component';
import { DialogTitleBarComponent } from './dialog-titlebar.component';
/**
 * Represents the Kendo UI Dialog component for Angular.
 */
export var DialogComponent = (function () {
    function DialogComponent(_elRef, _renderer) {
        this._elRef = _elRef;
        this._renderer = _renderer;
        /**
         * Fires when the user clicks on the **Close** button of the Dialog.
         */
        this.action = new EventEmitter();
        /**
         * Fires when the user clicks on the **Close** button of the Dialog.
         */
        this.close = new EventEmitter();
    }
    DialogComponent.prototype.ngAfterContentInit = function () {
        this.bubble('close', this.titlebarContent);
    };
    DialogComponent.prototype.ngAfterViewInit = function () {
        this.bubble('close', this.titlebarView);
        this.bubble('action', this.actionsView);
    };
    DialogComponent.prototype.ngOnInit = function () {
        this._renderer.setElementAttribute(this._elRef.nativeElement, 'title', null);
    };
    Object.defineProperty(DialogComponent.prototype, "wrapperClass", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    DialogComponent.prototype.bubble = function (eventName, component) {
        var _this = this;
        if (component) {
            component[eventName].subscribe(function (e) { return _this[eventName].emit(e); });
        }
    };
    DialogComponent.decorators = [
        { type: Component, args: [{
                    animations: [
                        trigger('overlayAppear', [
                            state('in', style({ opacity: 1 })),
                            transition('void => *', [
                                style({ opacity: .1 }),
                                animate('.3s cubic-bezier(.2, .6, .4, 1)')
                            ])
                        ]),
                        trigger('dialogSlideInAppear', [
                            state('in', style({ transform: 'translate(-50%, -50%)' })),
                            transition('void => *', [
                                style({ transform: 'translate(-50%, -60%)' }),
                                animate('.3s cubic-bezier(.2, 1, .2, 1)')
                            ])
                        ])
                    ],
                    exportAs: 'kendoDialog',
                    selector: 'kendo-dialog',
                    template: "\n    <div class=\"k-overlay\" @overlayAppear></div>\n\n    <div class=\"k-widget k-window k-dialog k-centered k-dialog-centered\" @dialogSlideInAppear>\n\n      <kendo-dialog-titlebar *ngIf=\"title\">{{title}}</kendo-dialog-titlebar>\n      <ng-content select=\"kendo-dialog-titlebar\" *ngIf=\"!title\"></ng-content>\n\n      <div class=\"k-content k-window-content k-dialog-content\">\n        <ng-content *ngIf=\"!contentTemplate\"></ng-content>\n        <template [ngTemplateOutlet]=\"contentTemplate\" *ngIf=\"contentTemplate\"></template>\n      </div>\n\n      <ng-content select=\"kendo-dialog-actions\" *ngIf=\"!actions\"></ng-content>\n      <kendo-dialog-actions [actions]=\"actions\" *ngIf=\"actions\"></kendo-dialog-actions>\n\n    </div>\n  "
                },] },
    ];
    /** @nocollapse */
    DialogComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer, },
    ]; };
    DialogComponent.propDecorators = {
        'title': [{ type: Input },],
        'actions': [{ type: Input },],
        'action': [{ type: Output },],
        'close': [{ type: Output },],
        'titlebarContent': [{ type: ContentChild, args: [DialogTitleBarComponent,] },],
        'titlebarView': [{ type: ViewChild, args: [DialogTitleBarComponent,] },],
        'actionsView': [{ type: ViewChild, args: [DialogActionsComponent,] },],
        'wrapperClass': [{ type: HostBinding, args: ['class.k-dialog-wrapper',] },],
    };
    return DialogComponent;
}());
