// @progress/kendo-angular-grid/dist/es/pager.component.js
import { Component, Input, EventEmitter, Output, HostBinding } from '@angular/core';
import { normalize } from './pager-settings';
import { LocalizationService } from '@progress/kendo-angular-l10n';
/**
 * @hidden
 */
export var PagerComponent = (function () {
    function PagerComponent(localization) {
        this.localization = localization;
        this.total = 0;
        this.skip = 1;
        this.pageChange = new EventEmitter();
        this.settings = normalize({});
    }
    Object.defineProperty(PagerComponent.prototype, "options", {
        set: function (value) {
            this.settings = normalize(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagerComponent.prototype, "pagerWrapClass", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagerComponent.prototype, "gridPagerClass", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagerComponent.prototype, "widgetClass", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagerComponent.prototype, "prevButtonsDisabled", {
        get: function () {
            return this.currentPage === 1 || !this.total;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagerComponent.prototype, "nextButtonsDisabled", {
        get: function () {
            return this.currentPage === this.totalPages || !this.total;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagerComponent.prototype, "showInitialPageSize", {
        get: function () {
            var _this = this;
            return this.settings.pageSizes
                .filter(function (num) { return num === Number(_this.pageSize); })
                .length === 0;
        },
        enumerable: true,
        configurable: true
    });
    PagerComponent.prototype.pageSizeChange = function (value) {
        this.pageSize = parseInt(value, 10);
        this.changePage(0);
    };
    PagerComponent.prototype.changePage = function (page) {
        this.pageChange.emit({ skip: page * this.pageSize, take: this.pageSize });
        return false;
    };
    Object.defineProperty(PagerComponent.prototype, "currentPage", {
        get: function () {
            return Math.floor((this.skip || 0) / this.pageSize) + 1;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagerComponent.prototype, "totalPages", {
        get: function () {
            return Math.ceil((this.total || 0) / this.pageSize);
        },
        enumerable: true,
        configurable: true
    });
    PagerComponent.prototype.onInputChange = function (value) {
        var page = parseInt(value, 10);
        if (isNaN(page) || page < 1 || page > this.totalPages) {
            page = this.currentPage;
        }
        this.changePage(page - 1);
    };
    Object.defineProperty(PagerComponent.prototype, "buttons", {
        get: function () {
            var result = [];
            for (var idx = this.start; idx <= this.end; idx++) {
                result.push(idx);
            }
            return result;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagerComponent.prototype, "maxItems", {
        get: function () {
            return Math.min(this.currentPage * this.pageSize, this.total);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagerComponent.prototype, "currentPageText", {
        get: function () {
            return this.total ?
                (this.currentPage - 1) * this.pageSize + 1 :
                0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagerComponent.prototype, "end", {
        get: function () {
            return Math.min((this.start + this.settings.buttonCount) - 1, this.totalPages);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagerComponent.prototype, "start", {
        get: function () {
            var page = this.currentPage;
            var buttonCount = this.settings.buttonCount;
            if (page > buttonCount) {
                var reminder = (page % buttonCount);
                return (reminder === 0) ? (page - buttonCount) + 1 : (page - reminder) + 1;
            }
            return 1;
        },
        enumerable: true,
        configurable: true
    });
    PagerComponent.prototype.textFor = function (key) {
        return this.localization.get(key);
    };
    PagerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'kendo-pager',
                    template: "\n        <a *ngIf=\"settings.previousNext\"\n            href=\"#\"\n            [title]=\"textFor('pagerFirstPage')\"\n            (click)=\"currentPage !== 1 ? changePage(0) : false\"\n            [ngClass]=\"{\n                'k-link': true,\n                'k-pager-nav': true,\n                'k-state-disabled': prevButtonsDisabled,\n                'k-pager-first': true\n            }\">\n            <span [attr.aria-label]=\"textFor('pagerFirstPage')\"\n                [ngClass]=\"{\n                    'k-icon':true,\n                    'k-i-seek-w': true\n                }\">\n            </span>\n        </a>\n        <a *ngIf=\"settings.previousNext\"\n            href=\"#\"\n            [title]=\"textFor('pagerPreviousPage')\"\n            (click)=\"currentPage !== 1 ? changePage(currentPage-2) : false\"\n            [ngClass]=\"{\n                'k-link': true,\n                'k-pager-nav': true,\n                'k-state-disabled': prevButtonsDisabled,\n                '': true\n            }\">\n            <span [attr.aria-label]=\"textFor('pagerPreviousPage')\"\n                [ngClass]=\"{\n                    'k-icon':true,\n                    'k-i-arrow-w': true\n                }\">\n            </span>\n        </a>\n        <ul [ngClass]=\"{'k-pager-numbers': true, 'k-reset': true}\" *ngIf=\"settings.type === 'numeric'\">\n            <li *ngIf=\"start > 1\"><a class=\"k-link\" href=\"#\" (click)=\"changePage(start - 2)\">...</a></li>\n            <li *ngFor=\"let num of buttons\">\n                <a href=\"#\"\n                    [ngClass]=\"{'k-link': true, 'k-state-selected':currentPage == num}\"\n                    (click)=\"changePage(num - 1)\">\n                    {{num}}\n                </a>\n            </li>\n            <li *ngIf=\"end < totalPages\"><a class=\"k-link\" href=\"#\" (click)=\"changePage(end)\">...</a></li>\n        </ul>\n        <span *ngIf=\"settings.type === 'input'\" [ngClass]=\"{'k-pager-input': true, 'k-label': true}\">\n            {{textFor('pagerPage')}}\n            <input [class.k-textbox]=\"true\" [value]=\"currentPage\" #inputPager (change)=\"onInputChange(inputPager.value)\" />\n            {{textFor('pagerOf')}} {{totalPages}}\n        </span>\n        <a *ngIf=\"settings.previousNext\"\n            href=\"#\"\n            [title]=\"textFor('pagerNextPage')\"\n            (click)=\"currentPage !== totalPages ? changePage(currentPage) : false\"\n            [ngClass]=\"{\n                'k-link': true,\n                'k-pager-nav': true,\n                'k-state-disabled': nextButtonsDisabled,\n                '': true\n            }\">\n            <span [attr.aria-label]=\"textFor('pagerNextPage')\"\n                [ngClass]=\"{\n                    'k-icon':true,\n                    'k-i-arrow-e': true\n                }\">\n            </span>\n        </a>\n        <a *ngIf=\"settings.previousNext\"\n            href=\"#\"\n            [title]=\"textFor('pagerLastPage')\"\n            (click)=\"currentPage !== totalPages ? changePage(totalPages-1) : false\"\n            [ngClass]=\"{\n                'k-link': true,\n                'k-pager-nav': true,\n                'k-state-disabled': nextButtonsDisabled,\n                'k-pager-last': true\n            }\">\n            <span [attr.aria-label]=\"textFor('pagerLastPage')\"\n                [ngClass]=\"{\n                    'k-icon':true,\n                    'k-i-seek-e': true\n                }\">\n            </span>\n        </a>\n        <span *ngIf='settings.info' [ngClass]=\"{'k-pager-info': true, 'k-label': true}\">\n            {{currentPageText}} - {{maxItems}} {{textFor('pagerOf')}} {{total}} {{textFor('pagerItems')}}\n        </span>\n        <span *ngIf=\"settings.pageSizes\" [ngClass]=\"{'k-pager-sizes': true, 'k-label': true}\">\n            <select #pageSizes (change)=\"pageSizeChange(pageSizes.value)\">\n                <option *ngIf=\"showInitialPageSize\" [value]=\"pageSize\">{{pageSize}}</option>\n                <option *ngFor=\"let page of settings.pageSizes\" [value]=\"page\" [selected]=\"page === pageSize ? true : undefined\">\n                    {{page}}\n                </option>\n            </select>\n            {{textFor('pagerItemsPerPage')}}\n        </span>\n  "
                },] },
    ];
    /** @nocollapse */
    PagerComponent.ctorParameters = function () { return [
        { type: LocalizationService, },
    ]; };
    PagerComponent.propDecorators = {
        'total': [{ type: Input },],
        'skip': [{ type: Input },],
        'pageSize': [{ type: Input },],
        'options': [{ type: Input },],
        'pageChange': [{ type: Output },],
        'pagerWrapClass': [{ type: HostBinding, args: ['class.k-pager-wrap',] },],
        'gridPagerClass': [{ type: HostBinding, args: ['class.k-grid-pager',] },],
        'widgetClass': [{ type: HostBinding, args: ['class.k-widget',] },],
    };
    return PagerComponent;
}());
