// @progress/kendo-angular-grid/dist/es/grouping/group-panel.component.js
import { Component, HostBinding, Input, ElementRef, Output, EventEmitter, ChangeDetectionStrategy, ViewChildren } from '@angular/core';
import { GroupConnectionService, GroupDragService } from './group-connection.service';
import { DraggableDirective } from '../draggable.directive';
import { observe } from '../utils';
import { LocalizationService } from '@progress/kendo-angular-l10n';
/**
 * @hidden
 */
export var GroupPanelComponent = (function () {
    function GroupPanelComponent(connection, draggableService, element, localization) {
        var _this = this;
        this.connection = connection;
        this.draggableService = draggableService;
        this.localization = localization;
        this.change = new EventEmitter();
        this.groups = [];
        connection
            .register(element.nativeElement)
            .subscribe(function (_a) {
            var field = _a.field, idx = _a.idx;
            return _this.insert(field, idx);
        });
    }
    Object.defineProperty(GroupPanelComponent.prototype, "groupHeaderClass", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GroupPanelComponent.prototype, "text", {
        get: function () {
            return this.emptyText ? this.emptyText : this.localization.get('groupPanelEmpty');
        },
        set: function (value) {
            this.emptyText = value;
        },
        enumerable: true,
        configurable: true
    });
    GroupPanelComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        observe(this.dropTargets)
            .subscribe(function (items) {
            return _this.connection.registerItems(items.map(function (x) { return x.nativeElement; }));
        });
        observe(this.draggables)
            .subscribe(function (items) {
            return _this.draggableService.connect(items.toArray());
        });
    };
    GroupPanelComponent.prototype.directionChange = function (group) {
        var index = this.groups.findIndex(function (x) { return x.field === group.field; });
        var groups = this.groups.slice(0, index).concat([group], this.groups.slice(index + 1));
        this.change.emit(groups);
    };
    GroupPanelComponent.prototype.insert = function (field, index) {
        var groups = this.groups.filter(function (x) { return x.field !== field; });
        if (groups.length || this.groups.length === 0) {
            this.change.emit(groups.slice(0, index).concat([{ field: field }], groups.slice(index)));
        }
    };
    GroupPanelComponent.prototype.remove = function (group) {
        this.change.emit(this.groups.filter(function (x) { return x.field !== group.field; }));
    };
    GroupPanelComponent.decorators = [
        { type: Component, args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [GroupDragService],
                    selector: 'kendo-grid-group-panel',
                    template: "\n    <template [ngIf]=\"groups.length === 0\">\n        {{ text }}\n    </template>\n    <div *ngFor=\"let group of groups\"\n        kendoGroupIndicator\n        [kendoGridDraggable]=\"group\"\n        [group]=\"group\"\n        (directionChange)=\"directionChange($event)\"\n        (remove)=\"remove($event)\">\n    </div>\n    "
                },] },
    ];
    /** @nocollapse */
    GroupPanelComponent.ctorParameters = function () { return [
        { type: GroupConnectionService, },
        { type: GroupDragService, },
        { type: ElementRef, },
        { type: LocalizationService, },
    ]; };
    GroupPanelComponent.propDecorators = {
        'change': [{ type: Output },],
        'groupHeaderClass': [{ type: HostBinding, args: ["class.k-grouping-header",] },],
        'text': [{ type: Input },],
        'groups': [{ type: Input },],
        'draggables': [{ type: ViewChildren, args: [DraggableDirective,] },],
        'dropTargets': [{ type: ViewChildren, args: [DraggableDirective, { read: ElementRef },] },],
    };
    return GroupPanelComponent;
}());
